gammapy/datasets/map.py

```diff
@@ -212,9 +212,27 @@ def models(self, models):
     @property
     def evaluators(self):
         """Model evaluators"""
-        # this call is needed to trigger the setup of the evaluators
-        if not self._evaluators:
-            self.npred()
+
+        if self.models:
+            hex_ids = []
+            for model in self.models:
+                evaluator = self._evaluators.get(model.hex_id)
+                hex_ids.append(model.hex_id)
+                if evaluator is None:
+                    evaluator = MapEvaluator(
+                        model=model, evaluation_mode=self.evaluation_mode, gti=self.gti
+                    )
+                    self._evaluators[model.hex_id] = evaluator
```

Comment:
We can actually use the model directly as a key here, like `self._evaluators[model]`, this results in the same behaviour but will be simpler. It also avoids to introduce the `model.hex_id` property. Maybe it's also an option then use set differences likes:

```python3
evaluators = set(["a", "b", "c"])
models = set(["b", "d", "f"])

# evaluators to delete    
for key in evaluators.difference(models):
    print(key)
```